at_lookup/lib/src/at_lookup_impl.dart

```diff
@@ -350,63 +350,77 @@ class AtLookupImpl implements AtLookUp {
-    }
-    fromResponse = fromResponse.trim().replaceAll('data:', '');
-    logger.finer('fromResponse $fromResponse');
-    var key = RSAPrivateKey.fromString(privateKey);
-    var sha256signature =
-        key.createSHA256Signature(utf8.encode(fromResponse) as Uint8List);
-    var signature = base64Encode(sha256signature);
-    logger.finer('Sending command pkam:$signature');
-    await _sendCommand('pkam:$signature\n');
-    var pkamResponse = await messageListener.read();
-    if (pkamResponse == 'data:success') {
-      logger.info('auth success');
-      _isPkamAuthenticated = true;
-    } else {
-      throw UnAuthenticatedException(
-          'Failed connecting to $_currentAtSign. $pkamResponse');
+    try {
+      pkamAuthenticationMutex.acquire();
+      await _sendCommand('from:$_currentAtSign\n');
+      var fromResponse = await (messageListener.read());
+      logger.finer('from result:$fromResponse');
+      if (fromResponse.isEmpty) {
+        return false;
+      }
+      fromResponse = fromResponse.trim().replaceAll('data:', '');
+      logger.finer('fromResponse $fromResponse');
+      var key = RSAPrivateKey.fromString(privateKey);
+      var sha256signature =
+          key.createSHA256Signature(utf8.encode(fromResponse) as Uint8List);
+      var signature = base64Encode(sha256signature);
+      logger.finer('Sending command pkam:$signature');
+      await _sendCommand('pkam:$signature\n');
+      var pkamResponse = await messageListener.read();
+      if (pkamResponse == 'data:success') {
+        logger.info('auth success');
+        _isPkamAuthenticated = true;
+      } else {
+        throw UnAuthenticatedException(
+            'Failed connecting to $_currentAtSign. $pkamResponse');
+      }
+      return _isPkamAuthenticated;
+    } finally {
+      pkamAuthenticationMutex.release();
     }
-    return _isPkamAuthenticated;
   }
 
+  Mutex cramAuthenticationMutex = Mutex();
+
   /// Generates digest using from verb response and [secret] and performs a CRAM authentication to
   /// secondary server
   Future<bool> authenticate_cram(var secret) async {
     secret ??= cramSecret;
     if (secret == null) {
       throw UnAuthenticatedException('Cram secret not passed');
     }
-    await _sendCommand('from:$_currentAtSign\n');
-    var fromResponse = await messageListener.read();
-    logger.info('from result:$fromResponse');
-    if (fromResponse.isEmpty) {
-      return false;
-    }
-    fromResponse = fromResponse.trim().replaceAll('data:', '');
-    var digestInput = '$secret$fromResponse';
-    var bytes = utf8.encode(digestInput);
-    var digest = sha512.convert(bytes);
-    await _sendCommand('cram:$digest\n');
-    var cramResponse = await messageListener.read();
-    if (cramResponse == 'data:success') {
-      logger.info('auth success');
-      _isCramAuthenticated = true;
-    } else {
-      throw UnAuthenticatedException('Auth failed');
+    try {
+      cramAuthenticationMutex.acquire();
```

Comment:
Wrap the next block of code in an if condition
```
if (!_isCramAuthenticated) {
    await _sendCommand ..... etc
}
```

Reply:
The next block of code is moved into IF condition.